poly/promise/promise.go

```diff
@@ -0,0 +1,1355 @@
+		}
+	}
+	return nil
+}
+
+/* Marshals a Promise struct using an io.Writer
+ *
+ * Arguments
+ *    w = the writer to use for marshalling
+ *
+ * Returns
+ *   The number of bytes written
+ *   The error status of the write (nil if no errors)
+ */
+func (p *Promise) MarshalTo(w io.Writer) (int, error) {
+	buf, err := p.MarshalBinary()
+	if err != nil {
+		return 0, err
+	}
+	return w.Write(buf)
+}
+
+/* Unmarshals a Promise struct using an io.Reader
+ *
+ * Arguments
+ *    r = the reader to use for unmarshalling
+ *
+ * Returns
+ *   The number of bytes read
+ *   The error status of the read (nil if no errors)
+ */
+func (p *Promise) UnmarshalFrom(r io.Reader) (int, error) {
+	// Retrieve p.n and p.t and then initialize p.PubPoly
+	buf := make([]byte, 2*uint32Size)
+	n, err := io.ReadFull(r, buf)
+	if err != nil {
+		return n, err
+	}
+	p.n = int(binary.LittleEndian.Uint32(buf))
+	p.t = int(binary.LittleEndian.Uint32(buf[uint32Size:]))
+	p.pubPoly = poly.PubPoly{}
+	p.pubPoly.Init(p.shareSuite, p.t, nil)
+
+	// MarshalSize can now be used to construct the final buffer. Copy
+	// the contents into the buffer and unmarshal.
+	finalBuf := make([]byte, p.MarshalSize())
+	copy(finalBuf, buf)
+	m, err := io.ReadFull(r, finalBuf[n:])
+	if err != nil {
+		return n + m, err
+	}
+	return n + m, p.UnmarshalBinary(finalBuf)
+}
+
+/* Returns a string representation of the Promise for easy debugging
+ *
+ * Returns
+ *   The Promise's string representation
+ */
+func (p *Promise) String() string {
+	s := "{Promise:\n"
+	s += "ShareSuite => " + p.shareSuite.String() + ",\n"
+	s += "KeySuite => " + p.keySuite.String() + ",\n"
+	s += "t => " + strconv.Itoa(p.t) + ",\n"
+	s += "r => " + strconv.Itoa(p.r) + ",\n"
+	s += "n => " + strconv.Itoa(p.n) + ",\n"
+	s += "Public Key => " + p.pubKey.String() + ",\n"
+	s += "Public Polynomial => " + p.pubPoly.String() + ",\n"
+	insurers := ""
+	secrets := ""
+	for i := 0; i < p.n; i++ {
+		insurers += p.insurers[i].String() + ",\n"
+		secrets += p.secrets[i].String() + ",\n"
+	}
+	s += "Insurers =>\n[" + insurers + "],\n"
+	s += "Secrets =>\n[" + secrets + "]\n"
+	s += "}\n"
+	return s
+}
+
+/* The PromiseState struct is responsible for maintaining state about Promise
+ * structs. It consists of three main pieces:
+ *
+ *    1. The promise itself, which should be treated like an immutable object
+ *    2. The shared secrets the server has recovered so far
+ *    3. A list of signatures from insurers cerifying the promise
+ *
+ * Each server should have one PromiseState per Promise
+ *
+ * Note to users of this code:
+ *
+ *    To add a share to PriShares, do:
+ *
+ *       p.PriShares.SetShare(index, share)
+ *
+ *    To reconstruct the secret, do:
+ *
+ *       p.PriShares.Secret()
+ *
+ *    Be warned that Secret will panic unless there are enough shares to
+ *    reconstruct the secret. (See poly/sharing.go for more info)
+ *
+ * TODO Consider if it is worth adding a String function
+ */
+type PromiseState struct {
```

Comment:
Since this is in a 'promise' package, just name this 'State' instead of 'PromiseState': otherwise external code will be referring to this as 'promise.PromiseState' which is long and redundant.  Same with all the other Promise-prefixed public types etc.